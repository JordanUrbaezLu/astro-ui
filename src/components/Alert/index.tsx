/* istanbul ignore file */
export * from "./Alert";
