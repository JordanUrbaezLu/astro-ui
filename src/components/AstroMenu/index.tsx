export * from "./AstroMenu";
export * from "./AstroMenuItem";
