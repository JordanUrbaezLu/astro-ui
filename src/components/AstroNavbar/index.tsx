export * from "./AstroNavbar";
export * from "./AstroNavbarItem";
