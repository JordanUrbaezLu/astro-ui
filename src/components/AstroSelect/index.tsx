export * from "./AstroSelect";
export * from "./AstroSelectItem";
