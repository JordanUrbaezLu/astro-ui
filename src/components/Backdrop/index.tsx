/* istanbul ignore file */
export * from "./Backdrop";
