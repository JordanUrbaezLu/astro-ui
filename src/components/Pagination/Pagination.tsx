import classNames from "classnames";
import * as React from "react";
import { MdChevronLeft, MdChevronRight } from "react-icons/md";
import { IconButton } from "../IconButton";
import styles from "./Pagination.module.scss";

export type PaginationVariant = "primary" | "secondary";

export interface PaginationProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The current page to display for the Pagination
   */
  currentPage: number;
  /**
   * The callback fired when changing the page
   */
  onChangePage: (page: number) => void;
  /**
   * The amount of pages for the Pagination
   */
  pages: number;
  /**
   * The variant for the Pagination
   *
   * @default "secondary"
   */
  variant?: PaginationVariant;
}

export const Pagination: React.FC<PaginationProps> = (props) => {
  const {
    className,
    currentPage,
    onChangePage,
    pages,
    variant = "secondary",
    ...rest
  } = props;

  const decreasePage = () => {
    if (currentPage - 1 > 0) {
      onChangePage(currentPage - 1);
    }
  };

  const increasePage = () => {
    if (currentPage + 1 < pages + 1) {
      onChangePage(currentPage + 1);
    }
  };

  return (
    <nav
      className={classNames(styles.paginationContainer, className)}
      role="navigation"
      {...rest}
    >
      <IconButton
        className={classNames(styles.paginationIconButton)}
        content={<MdChevronLeft />}
        onClick={decreasePage}
        disabled={currentPage === 1}
        variant={variant}
      />
      <div className={styles.pagesButtonsContainer}>
        {currentPage - 2 > 0 ? (
          <IconButton
            className={classNames(
              styles.paginationIconButton,
              styles.ellipse
            )}
            tabIndex={-1}
            variant={variant}
          >
            ...
          </IconButton>
        ) : (
          <div style={{ height: "38px", width: "38px" }} />
        )}
        {Array.from({ length: pages }, (_, i) => i + 1).map(
          (page, index) => {
            if (
              (page === currentPage - 1 ||
                page === currentPage + 1) &&
              page !== 0
            ) {
              return (
                <IconButton
                  className={classNames(styles.paginationIconButton)}
                  key={index}
                  onClick={() => onChangePage(page)}
                  variant={variant}
                >
                  {page}
                </IconButton>
              );
            } else if (page === currentPage) {
              return (
                <IconButton
                  className={classNames(
                    page === 1 && styles.firstPage,
                    page === pages && styles.lastPage,
                    styles.paginationIconButton,
                    styles.currentPageIcon
                  )}
                  key={index}
                  tabIndex={-1}
                  variant={variant}
                >
                  {page}
                </IconButton>
              );
            } else {
              return null;
            }
          }
        )}
        {currentPage + 1 < pages ? (
          <IconButton
            className={classNames(
              styles.paginationIconButton,
              styles.ellipse
            )}
            tabIndex={-1}
            variant={variant}
          >
            ...
          </IconButton>
        ) : (
          <div style={{ height: "38px", width: "38px" }} />
        )}
      </div>
      <IconButton
        className={classNames(styles.paginationIconButton)}
        content={<MdChevronRight />}
        onClick={increasePage}
        disabled={currentPage === pages}
        variant={variant}
      />
    </nav>
  );
};
