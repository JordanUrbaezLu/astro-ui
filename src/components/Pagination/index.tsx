/* istanbul ignore file */
export * from "./Pagination";
