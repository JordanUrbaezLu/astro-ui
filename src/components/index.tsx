export * from "./Alert";
export * from "./Button";
export * from "./Checkbox";
export * from "./DatePicker";
export * from "./Link";
export * from "./Menu";
export * from "./Tooltip";
export * from "./Select";
export * from "./Switch";
export * from "./TextField";
