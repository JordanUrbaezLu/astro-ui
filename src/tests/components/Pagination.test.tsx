import { render, screen } from "@testing-library/react";
import { Pagination } from "../../components/Pagination";
import { axe } from "jest-axe";

describe("Pagination", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Pagination
        currentPage={2}
        pages={3}
        onChangePage={jest.fn()}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Pagination
        currentPage={2}
        pages={3}
        onChangePage={jest.fn()}
      />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Interaction", () => {
  test("Should decrease page on left chevron click.", () => {
    render(
      <Pagination
        currentPage={2}
        pages={3}
        onChangePage={jest.fn()}
      />
    );

    expect(screen.getAllByRole("button")[2]).toHaveTextContent("2");

    expect(screen.getAllByRole("button")[2]).toHaveClass(
      "currentPageIcon"
    );

    screen.getAllByRole("button")[0].click();

    expect(screen.getAllByRole("button")[2]).toHaveTextContent("2");

    expect(screen.getAllByRole("button")[2]).toHaveClass(
      "currentPageIcon"
    );

    screen.getAllByRole("button")[4].click();

    expect(screen.getAllByRole("button")[2]).toHaveTextContent("2");

    expect(screen.getAllByRole("button")[2]).toHaveClass(
      "currentPageIcon"
    );
  });

  test("Should not trigger onChangePage on page button click.", () => {
    render(
      <Pagination
        currentPage={1}
        pages={1}
        onChangePage={jest.fn()}
      />
    );

    screen.getAllByRole("button")[0].click();

    screen.getAllByRole("button")[2].click();
  });

  test("Should trigger onChangePage on page button click.", () => {
    const onChangePage = jest.fn();
    render(
      <Pagination
        currentPage={5}
        pages={10}
        onChangePage={onChangePage}
      />
    );

    screen.getAllByRole("button")[2].click();

    expect(onChangePage).toHaveBeenCalledTimes(1);
  });
});
