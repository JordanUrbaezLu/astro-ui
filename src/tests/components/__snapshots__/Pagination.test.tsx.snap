// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Should render component correctly. 1`] = `
<nav
  class="paginationContainer"
  role="navigation"
>
  <button
    aria-label="Icon Button"
    class="paginationIconButton iconButton medium secondary"
  >
    <div
      class="container"
      role="presentation"
    />
    <svg
      fill="currentColor"
      height="22"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="22"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
      />
    </svg>
  </button>
  <div
    class="pagesButtonsContainer"
  >
    <div
      style="height: 38px; width: 38px;"
    />
    <button
      aria-label="Icon Button"
      class="paginationIconButton children iconButton medium secondary"
    >
      <div
        class="container"
        role="presentation"
      />
      1
    </button>
    <button
      aria-label="Icon Button"
      class="paginationIconButton currentPageIcon children iconButton medium secondary"
      tabindex="-1"
    >
      <div
        class="container"
        role="presentation"
      />
      2
    </button>
    <button
      aria-label="Icon Button"
      class="paginationIconButton children iconButton medium secondary"
    >
      <div
        class="container"
        role="presentation"
      />
      3
    </button>
    <div
      style="height: 38px; width: 38px;"
    />
  </div>
  <button
    aria-label="Icon Button"
    class="paginationIconButton iconButton medium secondary"
  >
    <div
      class="container"
      role="presentation"
    />
    <svg
      fill="currentColor"
      height="22"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="22"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </button>
</nav>
`;
